ilya-moskovtsev/nodejs-aws-be
# Task-9

Original Task: https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task9-bff/task.md

http://localhost:3000/api/product-service/products
http://ilya-moskovtsev-bff-api-dev.us-east-1.elasticbeanstalk.com/api/product-service/products
https://sy6ni7vd5i.execute-api.us-east-1.amazonaws.com/dev/products

http://localhost:3000/api/product-service/products/a7866d08-d8c3-4fcc-a3c7-57f31111972d
http://ilya-moskovtsev-bff-api-dev.us-east-1.elasticbeanstalk.com/api/product-service/products/a7866d08-d8c3-4fcc-a3c7-57f31111972d
https://sy6ni7vd5i.execute-api.us-east-1.amazonaws.com/dev/products/a7866d08-d8c3-4fcc-a3c7-57f31111972d

Product not found
http://localhost:3000/api/product-service/products/a7866d08-d8c3-4fcc-a3c7-57f311119721
http://ilya-moskovtsev-bff-api-dev.us-east-1.elasticbeanstalk.com/api/product-service/products/a7866d08-d8c3-4fcc-a3c7-57f311119721
https://sy6ni7vd5i.execute-api.us-east-1.amazonaws.com/dev/products/a7866d08-d8c3-4fcc-a3c7-57f311119721

```
curl --location --request POST 'http://localhost:3000/api/product-service/products' \
--header 'Content-Type: application/json' \
--data-raw '{
    "count": 1,
    "description": "bff 9",
    "price": 1,
    "title": "bff 9",
    "image_url": "https://images.unsplash.com/photo-1575709527142-a93ed587bb83?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=60",
    "image_title": "bff 9"
}'
```
```
curl --location --request POST 'http://ilya-moskovtsev-bff-api-dev.us-east-1.elasticbeanstalk.com/api/product-service/products' \
--header 'Content-Type: application/json' \
--data-raw '{
    "count": 1,
    "description": "bff 9",
    "price": 1,
    "title": "bff 9",
    "image_url": "https://images.unsplash.com/photo-1575709527142-a93ed587bb83?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=60",
    "image_title": "bff 9"
}'
```
```
curl --location --request POST 'https://sy6ni7vd5i.execute-api.us-east-1.amazonaws.com/dev/products' \
--header 'Content-Type: application/json' \
--data-raw '{
    "count": 1,
    "description": "bff 10",
    "price": 1,
    "title": "bff 10",
    "image_url": "https://images.unsplash.com/photo-1575709527142-a93ed587bb83?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=60",
    "image_title": "bff 10"
}'
```

http://localhost:3000/api/cart-service/profile/cart
http://ilya-moskovtsev-bff-api-dev.us-east-1.elasticbeanstalk.com/api/cart-service/profile/cart
http://ilya-moskovtsev-cart-api-cart-api-develop.us-east-1.elasticbeanstalk.com/api/profile/cart

```
curl --location --request PUT 'http://localhost:3000/api/cart-service/profile/cart' \
--header 'Content-Type: application/json' \
--data-raw '{
    "items": [
        {
            "product": {
                "count": 444,
                "description": "Boat Description 4",
                "id": "ecc67c83-fdd8-4d39-a79d-a02ad52d433a",
                "price": "444",
                "title": "Boat 4",
                "image_url": "https://source.unsplash.com/daily/?boat",
                "image_title": "Boat Description 4"
            },
            "count": 1
        },
        {
            "product": {
                "count": 555,
                "description": "Boat Description 5",
                "id": "a7866d08-d8c3-4fcc-a3c7-57f31111972d",
                "price": "555",
                "title": "Boat 5",
                "image_url": "https://source.unsplash.com/daily/?boat",
                "image_title": "Boat Description 5"
            },
            "count": 1
        }
    ]
}'
```
```
curl --location --request PUT 'http://ilya-moskovtsev-bff-api-dev.us-east-1.elasticbeanstalk.com/api/cart-service/profile/cart' \
--header 'Content-Type: application/json' \
--data-raw '{
    "items": [
        {
            "product": {
                "count": 444,
                "description": "Boat Description 4",
                "id": "ecc67c83-fdd8-4d39-a79d-a02ad52d433a",
                "price": "444",
                "title": "Boat 4",
                "image_url": "https://source.unsplash.com/daily/?boat",
                "image_title": "Boat Description 4"
            },
            "count": 1
        },
        {
            "product": {
                "count": 555,
                "description": "Boat Description 5",
                "id": "a7866d08-d8c3-4fcc-a3c7-57f31111972d",
                "price": "555",
                "title": "Boat 5",
                "image_url": "https://source.unsplash.com/daily/?boat",
                "image_title": "Boat Description 5"
            },
            "count": 1
        }
    ]
}'
```
```
curl --location --request PUT 'http://ilya-moskovtsev-cart-api-cart-api-develop.us-east-1.elasticbeanstalk.com/api/profile/cart' \
--header 'Content-Type: application/json' \
--data-raw '{
    "items": [
        {
            "product": {
                "count": 444,
                "description": "Boat Description 4",
                "id": "ecc67c83-fdd8-4d39-a79d-a02ad52d433a",
                "price": "444",
                "title": "Boat 4",
                "image_url": "https://source.unsplash.com/daily/?boat",
                "image_title": "Boat Description 4"
            },
            "count": 1
        },
        {
            "product": {
                "count": 555,
                "description": "Boat Description 5",
                "id": "a7866d08-d8c3-4fcc-a3c7-57f31111972d",
                "price": "555",
                "title": "Boat 5",
                "image_url": "https://source.unsplash.com/daily/?boat",
                "image_title": "Boat Description 5"
            },
            "count": 1
        }
    ]
}'
```

**Evaluation criteria (each mark includes previous mark criteria)**
3 (done) - A working and correct express application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the task 9.1
5 (done) - The bff-service should be deployed with Elastic Beanstalk. The bff-service call should be redirected to the appropriate service : product-service or CART. The response from the bff-service should be the same as if product-service or CART services were called directly.
**Additional (optional) tasks**
+1 (done) - Add a cache at the bff-service level for a request to the getProductsList function of the product-service. The cache should expire in 2 minutes.
How to test:
Get products list
Create new product
Get products list - result shouldn’t have new product
Wait more than 2 minutes
Get products list - result should have new product
+1 (done) - Use NestJS to create bff-service instead of express

7/7
